File: examples/rainfall/rainfall/RasterModule.js
var RasterModule = function (view, zoom, map_width, map_height) {
    // Create the map tag:
    var map_tag = "<div style='width:" + map_width + "px; height:" + map_height + "px;border:1px dotted' id='mapid'></div>"
    // Append it to body:
    var div = $(map_tag)[0]
    $('#elements').append(div)

    // Create Leaflet map and raster layers
    var Lmap = L.map('mapid', {zoomSnap: 0.1}).setView(view, zoom)
    var dummyUrl = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNgYAAAAAMAASsJTYQAAAAASUVORK5CYII='
    var rasterLayers = [L.imageOverlay(dummyUrl, [[0, 0], [0, 0]]).addTo(Lmap)]

    this.render = function (imageOverlay) {
        rasterLayers.forEach(function (layer) {
            layer.remove()
        })
        rasterLayers = []
        imageOverlay.urls.forEach(function (layer) {
            rasterLayers.push(L.imageOverlay(layer, imageOverlay.bounds).addTo(Lmap))
        })
        Lmap.fitBounds(imageOverlay.bounds)
    }

    this.reset = function () {
        rasterLayers.forEach(function (layer) {
            layer.remove()
        })
        rasterLayers = [L.imageOverlay(dummyUrl, [[0, 0], [0, 0]]).addTo(Lmap)]
    }
}

File: examples/rainfall/rainfall/RasterModule.py
from mesa.visualization.ModularVisualization import VisualizationElement
from folium.utilities import image_to_url


class RasterModule(VisualizationElement):
    package_includes = ["leaflet.js"]
    local_includes = ["rainfall/RasterModule.js"]

    def __init__(self, view=[0, 0], zoom=10, map_height=500, map_width=500):
        super().__init__()

        self.map_height = map_height
        self.map_width = map_width
        self.view = view
        new_element = "new RasterModule({}, {}, {}, {})"
        new_element = new_element.format(view, zoom, map_width, map_height)
        self.js_code = "elements.push(" + new_element + ");"

    def render(self, model):
        layers = {
            "rasters": [image_to_url(layer.values.transpose([1, 2, 0])) for layer in model.grid.layers],
            "bounds": model.grid.bounds
        }
        return layers
